src/common/base32.h

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) 2023, The Monero Project
+//
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+/**
+ * @file Encode/Decode using Jamtis base32 encoding
+ * 
+ * We use the alphabet "xmrbase32cdfghijknpqtuwy01456789"
+ * 
+ * This alphabet was selected for the following reasons:
+ *     1. To have a unique prefix that distinguishes the encoding from other variants of "base32"
+ *     2. To contain all digits 0-9, allowing numeric values to be encoded in a human readable form
+ *     3. To normalize the symbols o->0, l->1, v->u and z->2 for human transcription correction
+ *
+ * Hypens can be used to space base32 encoded strings, and are ignored during the decoding process.
+*/
+
+#pragma once
+
+#include <cstddef>
+#include <cstdint>
+#include <string>
+
+namespace base32
+{
+enum class Error: ssize_t
+{
+    invalid_char = -1,    // encountered invalid character when decoding
+    not_enough_space = -2 // not enough space in pre-allocated buffers
+};
+
+enum class Mode
+{
+    encoded_lossy, // when decoding, discard odd encoded LSB bits left at end of tail (default).
```

Comment:
`Mode::binary_lossy` in useful for encoding exact blocks of 5 bits so that the encoded base32 string isn't as long. For example, Jamtis address body sizes will be an odd number of bits long, not divisible by 8. Thus, we can make the encoded string one byte shorter since there's leftover bits in the binary that we aren't using.

Reply:
You select `binary_lossy` by passing it as the `mode` parameter in each function.

Reply:
I thought it was expected when divisible by 5-bits that no additional values would be appended. I guess not. But then how does someone select _lossless_ mode? There is no `enum` for it.

Reply:
Although it isn't explicit, almost all base32 libraries take the "encoded lossy" approach which preserves every bit in the raw data and discards extraneous encoded string bits, since that's the expected behavior 90% of the time. That's the default behavior for this code too, but now you have the option.

Reply:
> But then how does someone select lossless mode? 

You could make a lossless mode if and only if you forced the user to only encode raw data for which the byte length is divisible by 5, and decode encoded strings of which the length is divisible by 8.

Reply:
It could be added, although I don't know when that would be useful.